src/modules/0.0.22/iasql_functions/iasql.ts

```diff
@@ -1037,3 +1039,633 @@ export async function continueUpgrade(
   });
   await install([...modsToInstall.values()], dbId, dbUser, false, true);
 }
+
+export async function commit(dbId: string, dryRun: boolean, context: Context) {
+  const t1 = Date.now();
+  logger.info(`Applying commit to ${dbId}`);
+  const versionString = await TypeormWrapper.getVersionString(dbId);
+  const importedModules = await getImportedModules(dbId, versionString);
+  let orm: TypeormWrapper | null = null;
+  try {
+    orm = await TypeormWrapper.createConn(dbId);
+    context.orm = orm;
+    const newStartCommit = await insertCommit(orm, dryRun ? 'preview_start' : 'start');
+    context.startCommit = newStartCommit;
+    const previousStartCommit = await getPreviousStartCommit(orm);
+    const changesToCommit: IasqlAuditLog[] = await orm.find(IasqlAuditLog, {
+      order: { ts: 'DESC' },
+      where: {
+        changeType: Not(In([AuditLogChangeType.START_COMMIT, AuditLogChangeType.END_COMMIT])),
+        ts: previousStartCommit
+          ? Between(previousStartCommit.ts, newStartCommit.ts)
+          : LessThan(newStartCommit.ts),
+        user: Not(config.db.user),
+      },
+    });
+    const tablesWithChanges = [...new Set(changesToCommit.map(c => c.tableName))];
+    const changesByEntity: { [key: string]: any[] } = getChangesByEntity(changesToCommit);
+
+    const installedModulesNames = (await orm.find(IasqlModule)).map((m: any) => m.name);
+    const installedModules: ModuleInterface[] = (Object.values(importedModules) as ModuleInterface[]).filter(
+      mod => installedModulesNames.includes(`${mod.name}@${mod.version}`),
+    );
+
+    const modulesWithChanges: ModuleInterface[] = getModulesWithChanges(
+      importedModules,
+      installedModules,
+      installedModulesNames,
+      tablesWithChanges,
+      versionString,
+    );
+    const t2 = Date.now();
+    logger.info(`Setup took ${t2 - t1}ms`);
+
+    const toCreate: Crupde = {};
+    const toUpdate: Crupde = {};
+    const toReplace: Crupde = {}; // Not actually used in sync mode, at least right now
+    const toDelete: Crupde = {};
+    if (modulesWithChanges.length) {
+      const rootToLeafOrder: ModuleInterface[] = sortModules(modulesWithChanges, installedModulesNames);
+      await commitSync(
+        dbId,
+        rootToLeafOrder,
+        context,
+        toCreate,
+        toUpdate,
+        toReplace,
+        toDelete,
+        dryRun,
+        changesByEntity,
+      );
+      return await commitApply(
+        dbId,
+        rootToLeafOrder,
+        context,
+        toCreate,
+        toUpdate,
+        toReplace,
+        toDelete,
+        dryRun,
+        changesByEntity,
+      );
```

Comment:
Interesting that you're doing the remote sync first instead of the apply of the user-generated changes. What's your reasoning here?

Reply:
Since they are sequential actions, my idea was, on commit, to always make sure we have the DB updated before applying changes. Also, it is easier to identify which changes need to be synced and which ones need to be ignored during the `sync` phase since we know which changes need to be made in the cloud. The other way around could be trickier to apply the filters.